Index: src/storage_conf.c
===================================================================
RCS file: /data/cvs/libvirt/src/storage_conf.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -r1.3 -r1.4
--- a/src/storage_conf.c	27 Feb 2008 10:37:19 -0000	1.3
+++ b/src/storage_conf.c	28 Mar 2008 17:56:44 -0000	1.4
@@ -479,7 +479,7 @@
     }
     if ((options->flags & VIR_STORAGE_BACKEND_POOL_SOURCE_DIR) &&
         def->source.dir &&
-        virBufferVSprintf(buf,"    <directory path='%s'/>\n", def->source.dir) < 0)
+        virBufferVSprintf(buf,"    <dir path='%s'/>\n", def->source.dir) < 0)
         goto no_memory;
     if ((options->flags & VIR_STORAGE_BACKEND_POOL_SOURCE_ADAPTER) &&
         def->source.adapter &&
