diff --git a/src/Makefile.am b/src/Makefile.am
index 9b54679..099f21a 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1275,6 +1275,7 @@ endif
 if WITH_LIBXL
 	$(MKDIR_P) "$(DESTDIR)$(localstatedir)/lib/libvirt/libxl"
 	$(MKDIR_P) "$(DESTDIR)$(localstatedir)/run/libvirt/libxl"
+	$(MKDIR_P) "$(DESTDIR)$(localstatedir)/log/libvirt/libxl"
 endif
 if WITH_UML
 	$(MKDIR_P) "$(DESTDIR)$(localstatedir)/lib/libvirt/uml"
