diff -rup libvirt-0.4.0.orig/src/xend_internal.c libvirt-0.4.0.new/src/xend_internal.c
--- libvirt-0.4.0.orig/src/xend_internal.c	2007-12-17 18:05:27.000000000 -0500
+++ libvirt-0.4.0.new/src/xend_internal.c	2008-01-18 21:13:30.000000000 -0500
@@ -1907,6 +1907,9 @@ sexpr_to_xend_node_info(struct sexpr *ro
     info->mhz = sexpr_int(root, "node/cpu_mhz");
     info->nodes = sexpr_int(root, "node/nr_nodes");
     info->sockets = sexpr_int(root, "node/sockets_per_node");
+    info->cores = sexpr_int(root, "node/cores_per_socket");
+    info->threads = sexpr_int(root, "node/threads_per_core");
+
     /* Xen 3.2.0 replaces sockets_per_node with 'nr_cpus'.
      * Old Xen calculated sockets_per_node using its internal
      * nr_cpus / (nodes*cores*threads), so fake it ourselves
@@ -1921,8 +1924,6 @@ sexpr_to_xend_node_info(struct sexpr *ro
         if (info->sockets == 0)
             info->sockets = 1;
     }
-    info->cores = sexpr_int(root, "node/cores_per_socket");
-    info->threads = sexpr_int(root, "node/threads_per_core");
     return (0);
 }
 
