Index: python/Makefile.am
===================================================================
RCS file: /data/cvs/libxen/python/Makefile.am,v
retrieving revision 1.13
diff -u -p -r1.13 Makefile.am
--- python/Makefile.am	5 Feb 2008 19:27:37 -0000	1.13
+++ python/Makefile.am	24 Sep 2008 13:21:06 -0000
@@ -50,7 +50,7 @@ GENERATED= libvirt-export.c \
 $(GENERATED): $(srcdir)/$(GENERATE) $(API_DESC)
 	$(PYTHON) $(srcdir)/$(GENERATE) $(srcdir)
 
-libvirt.py: $(srcdir)/libvir.py libvirtclass.py
+libvirt.py: $(srcdir)/libvir.py $(GENERATED)
 	cat $(srcdir)/libvir.py libvirtclass.py > $@-t
 	mv $@-t $@
 
*** python/Makefile.in.orig	2008-09-24 15:19:20.000000000 +0200
--- python/Makefile.in	2008-09-24 15:20:07.000000000 +0200
*************** uninstall-am: uninstall-local uninstall-
*** 1027,1033 ****
  @WITH_PYTHON_TRUE@$(GENERATED): $(srcdir)/$(GENERATE) $(API_DESC)
  @WITH_PYTHON_TRUE@	$(PYTHON) $(srcdir)/$(GENERATE) $(srcdir)
  
! @WITH_PYTHON_TRUE@libvirt.py: $(srcdir)/libvir.py libvirtclass.py
  @WITH_PYTHON_TRUE@	cat $(srcdir)/libvir.py libvirtclass.py > $@-t
  @WITH_PYTHON_TRUE@	mv $@-t $@
  
--- 1027,1033 ----
  @WITH_PYTHON_TRUE@$(GENERATED): $(srcdir)/$(GENERATE) $(API_DESC)
  @WITH_PYTHON_TRUE@	$(PYTHON) $(srcdir)/$(GENERATE) $(srcdir)
  
! @WITH_PYTHON_TRUE@libvirt.py: $(srcdir)/libvir.py $(GENERATED)
  @WITH_PYTHON_TRUE@	cat $(srcdir)/libvir.py libvirtclass.py > $@-t
  @WITH_PYTHON_TRUE@	mv $@-t $@
  
